### PR TITLE
DAML-LF: limit the instanciation of Proto en/de-coders of Cid, Val, and Nid.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
In this PR we make the protobuf encoders and decoder  for ContractId, Value and Nid sealed class within the `ValueCoder` and `TransactionCoder` object.
Hence we can control better how contract id are serialized.
    
The PR advances the state of #3830
   

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala [hunk 1]
```diff
@@ -25,56 +25,92 @@ import scala.collection.immutable.HashMap
 object TransactionCoder {
 
   import ValueCoder.{DecodeCid, EncodeCid, codecContractId}
-  type EncodeNid[Nid] = Nid => String
-  type EncodeVal[Cid <: Value.ContractId] =
-    VersionedValue[Cid] => Either[EncodeError, (ValueVersion, ValueOuterClass.VersionedValue)]
+
+  abstract class EncodeNid[Nid] private[lf] {
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala]:
```suggestion
  abstract class EncodeNid[-Nid] private[lf] {
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala [hunk 2]
```diff
@@ -25,56 +25,92 @@ import scala.collection.immutable.HashMap
...
+    def asString(id: Nid): String
+  }
+  abstract class DecodeNid[Nid] private[lf] {
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala]:
```suggestion
  abstract class DecodeNid[+Nid] private[lf] {
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala [hunk 3]
```diff
@@ -25,56 +25,92 @@ import scala.collection.immutable.HashMap
...
+    def fromString(s: String): Either[DecodeError, Nid]
+  }
+
+  val NidEncoder: EncodeNid[Value.NodeId] = new EncodeNid[Value.NodeId] {
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala]:
```suggestion
  val NidEncoder: EncodeNid[Value.NodeId] = _.index.toString
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala [hunk 4]
```diff
@@ -25,56 +25,92 @@ import scala.collection.immutable.HashMap
...
+    override def asString(id: Value.NodeId): String = id.index.toString
+  }
+
+  val NidDecoder: DecodeNid[Value.NodeId] = new DecodeNid[Value.NodeId] {
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala]:
```suggestion
  val NidDecoder: DecodeNid[Value.NodeId] = s => parseInt(s).fold(...)
```

and so on for the other instances.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala [hunk 5]
```diff
@@ -25,56 +25,92 @@ import scala.collection.immutable.HashMap
...
   /**
     * Encodes a contract instance with the help of the contractId encoding function
     * @param coinst the contract instance to be encoded
-    * @param encodeVal function to encode a value to protobuf
+    * @param encodeCid function to encode a cid to protobuf
     * @return protobuf wire format contract instance
     */
-  def encodeContractInstance[Cid <: Value.ContractId](
-      encodeVal: EncodeVal[Cid],
+  def encodeContractInstance[Cid](
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala]: 🆒

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala [hunk 6]
```diff
@@ -45,17 +45,63 @@ object ValueCoder {
...
+      case RelativeContractId(nid, _) => nid.index.toString -> true
+      case AbsoluteContractId(coid) => coid -> false
+    }
+  }
+  object AbsCidDecoder extends DecodeCid[AbsoluteContractId] {
+
+    override def fromString(s: String): Either[DecodeError, AbsoluteContractId] =
+      ContractIdString
+        .fromString(s)
+        .fold(
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala]: scalaz provides `bimap` on `Either` as well as `\/`.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala]: thanks.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala]: does not work with the `private[lf]`at line 29

**Pull request merged**